lib/winapi/sync.c

```diff
@@ -38,6 +40,135 @@ VOID LeaveCriticalSection (LPCRITICAL_SECTION lpCriticalSection)
     RtlLeaveCriticalSection(lpCriticalSection);
 }
 
+BOOL InitOnceBeginInitialize (LPINIT_ONCE lpInitOnce, DWORD dwFlags, PBOOL fPending, LPVOID *lpContext)
+{
+    assert(fPending != NULL);
+
+    if (dwFlags & INIT_ONCE_CHECK_ONLY) {
+        if (dwFlags & INIT_ONCE_ASYNC) {
+            *fPending = FALSE;
+            SetLastError(ERROR_INVALID_PARAMETER);
+            return FALSE;
+        }
+        if (((ULONG_PTR)lpInitOnce->Ptr & 3) != 3) {
```

Comment:
Where do these pointer checks come from? They aren't documented in the MSDN for example.

Reply:
They are, but implicitly:
> [...] If INIT_ONCE_CHECK_ONLY is specified and initialization has completed, the return value is TRUE.
> Otherwise, the return value is FALSE. To get extended error information, call GetLastError.

The `INIT_ONCE_CTX_RESERVED_BITS` lower bits of the pointer are used to keep track of state. Initialization hasn't completed yet if not both bits are set.

Reply:
Consider defining names (in the source file) for these 4 states and using a mask created from `INIT_ONCE_CTX_RESERVED_BITS` (or manually document that your masks and that value are related).

This will help making the code more self documenting, too (instead of having to describe the current state in a comment as currently done).

Reply:
Done.